lib/services/notifications.go

```diff
@@ -0,0 +1,323 @@
+	DeleteAllUserNotificationStatesForUser(ctx context.Context, username string) error
+	ListUserNotificationStates(ctx context.Context, username string, pageSize int, nextToken string) ([]notificationsv1.UserNotificationState, string, error)
+	UpsertUserLastSeenNotification(ctx context.Context, username string, ulsn *notificationsv1.UserLastSeenNotification) (notificationsv1.UserLastSeenNotification, error)
+	GetUserLastSeenNotification(ctx context.Context, username string) (notificationsv1.UserLastSeenNotification, error)
+	DeleteUserLastSeenNotification(ctx context.Context, username string) error
+}
+
+// ValidateNotification verifies that the necessary fields are configured for a notification object.
+func ValidateNotification(notification *notificationsv1.Notification) error {
+	if notification.SubKind == "" {
+		return trace.BadParameter("notification subkind is missing")
+	}
+
+	if notification.Spec == nil {
+		return trace.BadParameter("notification spec is missing")
+	}
+
+	if notification.Metadata == nil {
+		return trace.BadParameter("notification metadata is missing")
+	}
+
+	if notification.Metadata.Labels == nil {
+		return trace.BadParameter("notification metadata labels are missing")
+	}
+
+	return nil
+}
+
+// MarshalNotification marshals a Notification resource to JSON.
+func MarshalNotification(notification *notificationsv1.Notification, opts ...MarshalOption) ([]byte, error) {
+	if err := ValidateNotification(notification); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if !cfg.PreserveResourceID {
+		notification = proto.Clone(notification).(*notificationsv1.Notification)
+		//nolint:staticcheck // SA1019. Deprecated, but still needed.
+		notification.Metadata.Id = 0
+		notification.Metadata.Revision = ""
+	}
+	data, err := utils.FastMarshal(notification)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return data, nil
+}
+
+// UnmarshalNotification unmarshals a Notification resource from JSON.
+func UnmarshalNotification(data []byte, opts ...MarshalOption) (*notificationsv1.Notification, error) {
+	if len(data) == 0 {
+		return nil, trace.BadParameter("missing notification data")
+	}
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var obj notificationsv1.Notification
+	if err = utils.FastUnmarshal(data, &obj); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if cfg.ID != 0 {
+		//nolint:staticcheck // SA1019. Id is deprecated, but still needed.
+		obj.Metadata.Id = cfg.ID
+	}
+	if cfg.Revision != "" {
+		obj.Metadata.Revision = cfg.Revision
+	}
+	if !cfg.Expires.IsZero() {
+		obj.Metadata.Expires = timestamppb.New(cfg.Expires)
+	}
+
+	return &obj, nil
+}
+
+// ValidateGlobalNotification verifies that the necessary fields are configured for a global notification object.
+func ValidateGlobalNotification(globalNotification *notificationsv1.GlobalNotification) error {
```

Comment:
It doesn't need to be enforced because we don't use the metadata (and never should). All `GlobalNotification` does is wrap a `Notification` with a matcher, hence all the metadata for a global notification should be in the wrapped `Notification`'s metadata, which we do validate here. If anything, we should actually enforce that `globalNotification.metadata` should be nil, because if a client provides it, then it's doing something wrong (and likely meant to add metadata to the notification).